tokio/src/io/read_buf.rs

```diff
@@ -283,3 +280,15 @@ impl fmt::Debug for ReadBuf<'_> {
             .finish()
     }
 }
+
+fn slice_to_uninit_mut(slice: &mut [u8]) -> &mut [MaybeUninit<u8>] {
+    unsafe {slice::from_raw_parts_mut::<MaybeUninit<u8>>(slice.as_mut_ptr().cast(), slice.len())}
+}
+
+unsafe fn slice_assume_init(slice: &[MaybeUninit<u8>]) -> &[u8] {
```

Comment:
This is effectively the same operation as the unstable `MaybeUninit::slice_assume_init_ref`.  FWIW, that's implemented as
```
unsafe { &*(slice as *const [Self] as *const [T]) }
```
Maybe leave a comment to switch to std's impl if/when it's stable.  Same thing for `slice_assume_init_mut` <-> `MaybeUninit::slice_assume_init_mut`.

Reply:
I think I'll actually change it to this instead. Checking it out on godbolt showed that, when optimized, they're exactly the same (2 `mov` instructions), but without optimizations, the pointer manipulation is still 2 instructions while `from_raw_parts` is significantly larger. Using pointer manipulation should make debug builds faster, and save time optimizing (since it's already optimal)

Reply:
The PR is motivated with
>The layout of wide pointers is unspecified, so transmuteing them is not valid.

Pointer casts are a bit safer than `transmute` (since it's not just sizes of the src/dst that are checked) but I don't think this changes anything in terms of avoiding relying on fat pointer layout, so to speak.

Reply:
> but I don't think this changes anything in terms of avoiding relying on fat pointer layout

What is that in reference to, the method of conversion or the chance for optimization?

Reply:
Only pointing out that if you switch to using pointer casts, it "defeats" the stated motivation for this PR.  I think pointer casts are better than `transmute` (on previously mentioned grounds) but it doesn't side step the layout aspect, like `from_raw_parts{_mut}` would.

Reply:
How so? `as` casting is safe, even with pointers. If the layouts were to be different, it would result in instructions that converted between them (like casting `u32` to `f32`)

Reply:
Casting pointers is safe, yes, but not dereferencing them (and this code dereferences the cast to obtain a reference, i.e `&*` bit).  Compiler will not insert instructions to "convert" layouts.  This is the unsafety part: pointers can be freely cast around but once they're dereferenced (e.g. to obtain a reference), it's on the user to ensure the pointer is valid for the target type, such as layout.

Casting `u32` to `f32` via `as` is a scalar cast, it's different.

Reply:
The slice itself is valid, as `MaybeUninit` is `repr(transparent)`, it's the pointers themselves whose layout was the issue

Reply:
Let's take a step back.  The fat ptr is currently 2 words: (ptr_to_data, len).  Code prior to this PR `transmute`s this fat ptr to another fat ptr where the only thing changing is the `T` the data points to - it's effectively a `reinterpret_cast` of the (ptr_to_data, len).  As you mentioned, `MaybeUninit` is `repr(transparent)` so layout of `MaybeUninit<u8> == u8`.  What exact problem do you envision with using `transmute` here? If tomorrow the fat ptr layout changes to (len, ptr_to_data), `transmute` still works because it's only changing the `T` the slice points to.

Maybe I'm missing something here.

Reply:
`&[u8]` and `&[MaybeUninit<u8>]` aren't guaranteed to have the same layout (even though `[u8]` and `[MaybeUninit<u8>]` are). In theory, `&[u8]` could be (ptr, len) while `&[MaybeUninit<u8>]` could be (len, ptr). In practice, they are the same, but that isn't guaranteed and shouldn't be depended upon.

Reply:
>&[u8] and &[MaybeUninit<u8>] aren't guaranteed to have the same layout (even though [u8] and [MaybeUninit<u8>] are). In theory, &[u8] could be (ptr, len) while &[MaybeUninit<u8>] could be (len, ptr)

And you think compiler will insert layout "fixups" if casting between `*const [MaybeUninit<u8>]` and `*const [u8]` in such a scenario? Hmm, I think that's quite a leap.  Is that based on some information somewhere?

Reply:
> Hmm, I think that's quite a leap. Is that based on some information somewhere?

I guess no, I it isn't, just an educated guess, but I would find it very strange that if didn't since `as` conversion are safe to do, and casting wouldn't be *nearly* as useful if it wasn't valid to dereference it. And I don't see it as too much of a leap, `as` conversion that do work aren't unheard of, such as the aforementioned `u32` -> `f32`

I'll look for things to confirm or deny my point tomorrow, but it's getting rather late for me

Reply:
FWIW my understanding definitely agrees with @Cyborus04's here.

Conceptually the type `*const T` is `struct ConstPtr<T> { ptr: *const (), metadata: T::Metadata }` (where if T is a slice, then T::Metadata is usize).

Whether the ptr or metadata field comes first in memory is **not** a stable guarantee of the language, because that conceptual struct is not `repr(C)`. For different `T` it could be different whether ptr or metadata comes first &mdash; for example PGO might make a different choice based on the access patterns of each case. It would probably place the most frequently accessed field first, which may be the pointer in some slice types, and the length in other slice types.

So transmuting one `ConstPtr<[U]>` to `ConstPtr<[V]>` does not produce a meaningful result.

In contrast, `as` absolutely does produce a meaningful result. `p as *const [V]` is conceptually equivalent to `ConstPtr::<[V]> { ptr: p.ptr, metadata: p.metadata }` which does the runtime rearranging if the two happen to have those fields in opposite order.

Reply:
@dtolnay I buy that (and it’s a good way to think of it).  I suppose what you (and @Cyborus04) are saying is that the casting rules _would_ be expanded to cover such casts, including a mention of layout fixup (https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions currently says nothing about this).  I agree it’s reasonable, I just don’t see it spelled out anywhere (to avoid hypothesizing).

Thanks for the discussion!